### PR TITLE
DOC: try to be nicer about f2py.compile

### PR DESCRIPTION
**Pull request by HaoZeke**
The nicer version of #25193. Perhaps to be merged now and that can go in later.

### CODE REVIEW

**Comment by HaoZeke** [comment]
@mattip would it make sense to merge this for now and put #25193 when 2.0.0 is cut?

**Comment by mattip** [comment]
@melissawm thoughts?

File: numpy/f2py/__init__.py [hunk 1]
```diff
@@ -78,6 +80,13 @@ def compile(source,
     import tempfile
     import shlex
 
+    warnings.warn(
+        "f2py.compile() will be removed in NumPy 2.0.x.\n"
+        "Write a custom subprocess.run wrapper to f2py instead",
+        VisibleDeprecationWarning,
+        stacklevel=2,
```
Comment (melissawm) [inline numpy/f2py/__init__.py:87]: Maybe we want to include an example of this in the f2py docs.

**Review by melissawm — APPROVED** [review]
Looks reasonable to me, thanks @HaoZeke !

**Comment by rgommers** [comment]
> @mattip would it make sense to merge this for now and put #25193 when 2.0.0 is cut?

I'm not sure I understand this. If we want to remove `f2py.compile`, the time is either now, or otherwise a couple of releases into the future. Why not do it right now?

**Comment by HaoZeke** [comment]
> > @mattip would it make sense to merge this for now and put #25193 when 2.0.0 is cut?
> 
> I'm not sure I understand this. If we want to remove `f2py.compile`, the time is either now, or otherwise a couple of releases into the future. Why not do it right now?

I was wondering about the off-chance that it is being used and the downstream users would like some time to migrate away from it (though they really should so we can't keep it around forever). @seberg mentioned the release of 2.0 would be sometime early next year so there's still some time for people who update their versions regularly to get ahead of any issues they'd have.

File: numpy/f2py/__init__.py (hunk 1 above)
Comment (HaoZeke) [reply numpy/f2py/__init__.py:87]: +1 yup, though I think that'd be better suited to #25193.

**Comment by rgommers** [comment]
Well, we want to branch in a month and a bit I think, if everything goes well. And users of this - of which there will be few - are not likely to going to need it from a dev version. We have a lot of stuff to get in for 2.0, so if this is ready to go we should put it in.

**Comment by rgommers** [comment]
Okay, this is approved, so I'll hit the green button. Let's not leave anything to right before the branch cut though.

**Comment by rgommers** [comment]
Thank Rohit & reviewers!

**Pull request merged**